File: 프로그래머스/3/43164. 여행경로/여행경로.cpp
#include <bits/stdc++.h>

using namespace std;

vector<string> dfs(vector<vector<string>>& tickets, vector<string>& path, vector<bool>& visited, string cur, int tickets_size){
    path.push_back(cur);
    
    // 종료 조건
    if(path.size() == tickets_size + 1){
        return path;
    }
    
    for (int i = 0; i < tickets_size; i++){
        if (tickets[i][0] == cur && !visited[i]) {
            visited[i] = true;
            vector<string> result = dfs(tickets, path, visited, tickets[i][1], tickets_size);
            if(!result.empty()) return result;
            
            // 실패했다면 다시 복구하고 다음 i(다른 티켓)를 시도
            visited[i] = false; //백트래킹
        }
    }
    // 이 길로는 답이 안 나온다면 줬던 거 뺏고 돌아가기
    path.pop_back();
    
    // 실패 시 빈 벡터
    return {};
}


vector<string> solution(vector<vector<string>> tickets) {
    int len = tickets.size();
    vector<string> path;
    vector<bool> visited(len, false); //딱 맞춰서 초기화하기
    
    // 정렬해서 티켓 제공해서 사전순으로 정렬된 것부터 가도록 함. 
    sort(tickets.begin(), tickets.end());

    return dfs(tickets, path, visited, "ICN", len);
}

File: 프로그래머스/3/43164. 여행경로/여행경로.kt
class Solution {
    fun solution(tickets: Array<Array<String>>): Array<String> {
        val answer = mutableListOf<String>()
        
        // 정렬
        val sorted = tickets.sortedWith(
            compareBy<Array<String>> {it[0]}.thenBy{it[1]}
        )
        
        val route = mutableListOf("ICN")
        val visited = BooleanArray(tickets.size)
        
        fun dfs(route: MutableList<String>):MutableList<String>{
            if(route.size == sorted.size +1){
                for(r in route){
                    answer.add(r)
                }
                return answer //여기에서 반환을 하는데 왜 answer에 계속 덧붙여지는 거지???
            }
            
            val cur = route[route.size -1]
            
            //다음 경로 탐색
            for(i in sorted.indices){
                if(sorted[i][0] == cur && !visited[i]){
                    route.add(sorted[i][1])
                    visited[i] = true
                    
                    if (dfs(route).size == sorted.size +1){ //여기서도 반환을 해줘야되는구나!!
                        return answer
                    }
                    
                    // 그 경로가 맞지 않았다면
                    route.removeAt(route.size -1)
                    visited[i] = false
                }
            }
            
            return answer
        }
        
        return dfs(route).toTypedArray()
    }
}